src/main/scala/mesosphere/marathon/core/storage/impl/zk/ZkPersistenceStore.scala

```diff
@@ -0,0 +1,170 @@
+package mesosphere.marathon.core.storage.impl.zk
+
+import java.time.OffsetDateTime
+
+import akka.actor.Scheduler
+import akka.http.scaladsl.marshalling.Marshaller
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import akka.{ Done, NotUsed }
+import com.google.protobuf.MessageLite
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.marathon.core.storage.impl.BasePersistenceStore
+import mesosphere.marathon.core.storage.{ CategorizedKey, MarathonProto, MarathonState }
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.util.{ Retry, toRichFuture }
+import mesosphere.util.state.zk.{ Children, GetData, RichCuratorFramework }
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.NoNodeException
+import org.apache.zookeeper.data.Stat
+
+import scala.async.Async.{ async, await }
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+case class ZkId(category: String, id: String, version: Option[OffsetDateTime]) {
+  private val bucket = id.hashCode % 16
```

Comment:
All objects can be found in one of their categorized buckets (there are
only 16). This facilitates even distribution of the persisted data, but
yes, it makes it a bit harder if you're looking to debug.

In order to facilitate support, we agreed that a tool should be provided to
decode/search ZK for marathon data - plus, it can decode proto into a human
readable form.
On Tue, Jul 12, 2016 at 9:42 AM Ken Sipe <notifications@github.com> wrote:

> In
> src/main/scala/mesosphere/marathon/core/storage/impl/zk/ZkPersistenceStore.scala
> <https://github.com/mesosphere/marathon/pull/4062#discussion_r70474404>:
>
> > +import mesosphere.marathon.core.storage.{ CategorizedKey, MarathonProto, MarathonState }
> > +import mesosphere.marathon.metrics.Metrics
> > +import mesosphere.marathon.util.{ Retry, toRichFuture }
> > +import mesosphere.util.state.zk.{ Children, GetData, RichCuratorFramework }
> > +import org.apache.zookeeper.KeeperException
> > +import org.apache.zookeeper.KeeperException.NoNodeException
> > +import org.apache.zookeeper.data.Stat
> > +
> > +import scala.async.Async.{ async, await }
> > +import scala.collection.immutable.Seq
> > +import scala.concurrent.{ ExecutionContext, Future }
> > +import scala.util.control.NonFatal
> > +import scala.util.{ Failure, Success }
> > +
> > +case class ZkId(category: String, id: String, version: Option[OffsetDateTime]) {
> > +  private val bucket = id.hashCode % 16
>
> simply beautiful. I don't see a way to reverse this for debugging /
> support. It appears we save the full ID in the bucket. I just checked and
> Exhibitor doesn't provide a way of searching. Any thoughts here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4062/files/5ffaa7e4c7657fc609814fd932a7343acfb909f0#r70474404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAweZl2Vmjj2-QF5s_6dKKYf2-eoM3hhks5qU8PbgaJpZM4JDjhV>
> .
>